diff --git a/src/Emanote/Pandoc/Markdown/Parser.hs b/src/Emanote/Pandoc/Markdown/Parser.hs
index fb7ef43..254f275 100644
--- a/src/Emanote/Pandoc/Markdown/Parser.hs
+++ b/src/Emanote/Pandoc/Markdown/Parser.hs
@@ -39,6 +39,7 @@ parseMarkdown =
           CE.fencedDivSpec,
           CE.bracketedSpanSpec,
           CE.autolinkSpec,
+          CE.superscriptSpec,
           CM.defaultSyntaxSpec
         ]
     gfmExtensionsSansPipeTable =
