diff --git a/packages/backend/src/core/activitypub/ApInboxService.ts b/packages/backend/src/core/activitypub/ApInboxService.ts
index efef777fb..de6fba11a 100644
--- a/packages/backend/src/core/activitypub/ApInboxService.ts
+++ b/packages/backend/src/core/activitypub/ApInboxService.ts
@@ -302,6 +302,10 @@ export class ApInboxService {
 				return;
 			}
 
+			if (Array.isArray(activity.type) && (activity.type as string[]).some((t, i, a) => t == "Relay")) {
+				return;
+			}
+
 			this.logger.info(`Creating the (Re)Note: ${uri}`);
 
 			const activityAudience = await this.apAudienceService.parseAudience(actor, activity.to, activity.cc);
